### ISSUE TITLE
Macros, all parameters, parameter

### ISSUE THREAD

**Issue by AnthonyMuscio** [Jermolene/TiddlyWiki5#3750]
Folks, 
Is there a method already or could there be created a new core variable that allows us to reference all the parameters passed in a macro call with a single name? This feature is possible in a large number of scripting and batch programming languages including Javascript (I believe) but not in TiddlyWiki.

This would be very helpful in creating more advanced macros because it would be easier to pass on all parameters (without understanding there names and purpose) to another macro including existing core macros. Perhaps the first parameter will determine which secondary macro to call, it could even be the secondary macros name. 

For example
```
\define macroname(p1 p2 p3)
We can use $p1$, $p2$ and $p3$
\end
<<macroname "v1" "V2" parmname:"v3" newp>>
```
However if we wanted to pass all these parameters onto another macro it would be helpful if we could pass all the parameters on using a single substitution;
For example
```
\define macroname(p1 p2 p3)
<<newmacro $*$  newp>>
\end
<<macroname "v1" "V2" parmname:"v3">>
```
In this case $*$ would pass the following to newmacro
`"v1" "V2" parmname:"v3"  newp`

I do not know if this would be achievable or a TiddlyWiki X feature.

Regards
Tony

**Comment by AnthonyMuscio** [comment]
Mario,

I mearly want to pass all parameters on to the next macro. Of course the first macro can interogate the parameters, my point is it need not.

Regards
Tony

**Comment by Jermolene** [comment]
Hi @AnthonyMuscio I agree that the ability to pass along macro parameters in that way could be useful. As you suggest, an option on the `<$macrocall>` widget could request that all of the parameters to the current macro be passed, with any explicit parameters taking precedence.

However, the implementation wouldn't be straightforward with how things are structured at the moment. In particular, by the time we're parsing and rendering a macrocall widget there's no simple way to tell that the macrocall originated within another macro definition.

Another way of doing it would be similar to the way that we expose every macro parameter as a variable (quoted with double underscores). We could expose all of the parameters as a special variable `__params__`, but the trouble there is that there is no obvious representation to use to store the parameters within the variable.

**Comment by AnthonyMuscio** [comment]
Jeremy,

If such a __params__ where available it should contain parameters in the established formats ie; name:"value" name="value" and any others in use in tiddlywiki. If possible when a simple unnamed "value" is provided it should be given an incremental name eg p1 p2 

Perhaps choosing what the delimiter is would have value such as using "#" or "?" on a url would allow a parameter value pair returned, we can choose to use one, the other, or both the paramname or the value returned.

However I have realised we can already choose intentionally to pass all parameters within a single one, 
<$macrocall $name=macroname params="""string of parameters"""/> as below, however we do not have an easy way for the user to subsequently process the parameters. This can be passed on to a subsequent macro.

eg params="""tiddlername="tiddlerinstancename" class="classname" cost="10.00" "a value" """

I wonder if we "surfaced" parameter processing with a widget, the list widget and or a filter that accepted a "parameter string" and provided equivalent filter operators such as params<variable> hasparam[paramname] and getparam[paramname] (like fields[] has[fieldname] and get[fieldname] ). This would open up parameter processing to users without needing javascript solutions to be written. It would be here we can use the existing filter operators under "Order Operators" for first last rest etc...

Given above example
`params[$params$]` would list 
tiddlername class cost p1

hasparam[class] would list the param name if it exists
getparam[cost] would return the value for the paramname if it exists, including blank

**Basically, it is currently possible to generate parameter strings, but only existing macros and widgets can process them, there would be great value filling the gap such that solutions can be written that also can process parameter strings.**

**Comment by AnthonyMuscio** [comment]
Although I would like a core feature that given a `<<macro parm1 "parm1" "p 3">>` in the macro I could refer to $params$ which would contain `"""parm1 "parm1" "p 3"""` such that they can be passed on to a subsequent macro like `<<nextmacro $params$>>` I have also experimented with doing this myself.

`<<macro """parm1 "parm1" "p 3"""">>` such that the first parameter is equal to all parameters and I can then pass it on as a single parameter or break it down.

```
\define extractparameters(parameters delimiter)
All Parameters: "$parameters$" by "$delimiter$"<br>
<ol>
<$list filter="$parameters$" variable=eachparam>
   <$list filter="[<eachparam>splitbefore[$delimiter$]removesuffix[$delimiter$]]" variable=keyword>
   <$list filter="[<eachparam>removeprefix<keyword>removeprefix[$delimiter$]]" variable=value>
        <li><<keyword>>$delimiter$<<value>></li>
   </$list>
   </$list>
</$list>
</ol>
\end
<<extractparameters "test:value test2:value2" ":">>
```
Result:
**All Parameters: "test:value test2:value2" by ":"
test:value
test2:value2**

The truth is I would be nice to pass on butfirst($params$) and other possible combinations.

Regards
Tony

**Comment by Jermolene** [comment]
Hi @AnthonyMuscio 

> If such a __params__ where available it should contain parameters in the established formats ie; name:"value" name="value" and any others in use in tiddlywiki. If possible when a simple unnamed "value" is provided it should be given an incremental name eg p1 p2 

Using that format wouldn't work because it cannot express every legal attribute value: we have the old problem of how to delimit values that contain those delimiters. As I said above, we don't have an established format that will work here.

**Comment by Jermolene** [comment]
Just to be clear @AnthonyMuscio I think I understand what you're asking for, and I agree that it would be a useful capability. The issue is that I do not see a straightforward implementation.

**Comment by AnthonyMuscio** [comment]
Jeremy,

I am happy with my suggested 
`<<macro """parm1 "parm1" "p 3"""">>`

Now I am looking for or will build with string/maths operators a parser with the choice of delimiter eg = : 
For example to convert """parm1 "parm1" "p 3"""" to p1="parm1" p2="p 3" and p3="p 3" so I can pass it into a new-tiddler action/message such as

p1="parm1" p2="p 3" and p3="p 3"

Regards
Tony

**Comment by Jermolene** [comment]
Hi @AnthonyMuscio sorry for not being clear in my comment above, I'm afraid that suggestion is not robust enough to represent every legal attribute value.

**Comment by AnthonyMuscio** [comment]
Sorry Jeremy, In my ignorance I am not sure I understand.

I am somewhat content with building a way to pass values as a single parm for either passing on or parsing into time:value pairs to suit what every I am building. I think I can achive what I want with maths and string operators. 

What use cases /functions are you thinking of?

**Comment by Jermolene** [comment]
Hi @AnthonyMuscio the problem is how to represent the list of provided parameters so that we can manipulate it. We need a mechanism that can represent any set of macro parameter names and values, including, for example, ones that contain the characters we use to delimit them.

JSON would be an example of a format that would allow all of the names and values to be represented. In practice it would not be suitable both because of inefficiency and the lack of tools for manipulating blobs of JSON.

A change like this is right in the heart of how the TiddlyWiki core works and so it needs to be robustly engineered.

**Comment by AnthonyMuscio** [comment]
Jeremy,

The OT is can I pass all parameters in one macro onto another macro. the Simple answer to my initial question is **no,** there is no $parameters$ value I can reference. **The extended answer is Yes, We can emulate this.**

I will post the solution here shortly, for future reference, but then close this issue because an acceptable workaround is available.

By reviewing what I publish you may see that my requirements were simpler than expected, and it was possibly my fault for calling for;
> new core variable that allows us to reference all the parameters passed in a macro call with a single name?

Regards
Tony

**Comment by AnthonyMuscio** [comment]
Place the following in a tiddler to see all examples

The final example shows

- extracting the nth parameter, 
- iterating all parameters (to known limit) 
- extracting the value of parameter by name
- The one remaining bug **Note: p2 not correct**

Regards
Tony

```
\define macro1(parameters)
received = $parameters$<br>
Call macro2<br>
`<<macro2 $parameters$>> `<br>
<<macro2 $parameters$>> 
\end
\define macro2(p1 p2 p3)
<h3>Result</h3>
p1=$p1$, 
p2=$p2$, 
p3=$p3$, 
\end
\define macro-a(parameters)
received = $parameters$<br>
Using `set select=` to extract 2nd Parameter<br>
<$set name=param2 filter="""$parameters$""" select="1">
param2=<<param2>><br>
Using List and  `set select=` to extract all nth params (0 indexed)<br>
<$list filter="[range[0,2]]" variable=num>
<$set name=paramvalue filter="""$parameters$""" select=<<num>> >
<<num>>: <<paramvalue>><br>
</$set>
</$list>
''Note: p2 not correct''<br>
Using the parameter name to extract value from "p3" (2nd position) from received = "$parameters$"<br>
//Will use `contains[p3:]` to only show p3 in later version//<br>
<$list filter="[range[0,2]]" variable=num>
<$set name=paramvalue filter="""$parameters$ +[prefix[p3:]removeprefix[p3:"]removesuffix["]]""" select=<<num>> >
p3: = <<paramvalue>><br>
</$set>
</$list>

`<<macro2 $parameters$>> `<br>
<<macro2 $parameters$>> 
</$set>
\end
<hr>

;Test unnamed/undelimited parameters via 1st macro
://via macro1 selective delimiters//
`<<macro1 """ v1 "v 2" v3 """>>`<br>
<<macro1 """ v1 "v 2" v3 """>>
<hr>

;Test unnamed parameters via 1st macro
://via macro1, all delimited//
`<<macro1 """ "v1" "v 2" "v3" """>>`<br>
<<macro1 """ "v1" "v 2" "v3" """>>
<hr>

;Test named parameters via 1st macro
://via macro1, all delimited//
`<<macro1 """ p1:"v1"  p3:"v3"  p2:"v 2" """>>`<br>
<<macro1 """ p1:"v1"  p3:"v3"  p2:"v 2"  """>>
<hr>

;Test named parameters direct
://Direct to macro2//
`<<macro2  p1:"v1"  p3:"v3"  p2:"v 2" >> `<br>
<<macro2  p1:"v1" p3:"v3"  p2:"v 2"  >> 
<hr>

;Test extract parameters in 1st macro
://via macro-a, all delimited, to extract parameter 2, in macro-a//
`<<macro-a  """ p1:"v1"  p3:"v3"  p2:"v 2" """>> `<br>
<<macro-a  """ p1:"v1" p3:"v3"  p2:"v 2" """ >> 
```

**Comment by AnthonyMuscio** [comment]
Note the ability to extract say the value of p1 whilst still passing all parameters on to the second macro is p1 may be used as the second macro name, thus documenting a single macro which calls multiple "sub macros" as if they were "functions".

Regards
Tony

**Comment by AnthonyMuscio** [comment]
Yes, I acknowledge this in included within https://github.com/Jermolene/TiddlyWiki5/pull/6666

Thanks Jeremy
